src/Microsoft.ML.AutoML/API/ExperimentResults/CrossValidationExperimentResult.cs

```diff
@@ -36,5 +37,31 @@ public class CrossValidationExperimentResult<TMetrics>
             RunDetails = runDetails;
             BestRun = bestRun;
         }
+
+        #region IDisposable Support
+        private bool _disposed;
+
+        /// <summary>
+        /// Releases unmanaged Tensor objects in models stored in CrossValidationRunDetail instances
+        /// </summary>
+        /// <remarks>
+        /// Invocation of Dispose() is necessary to clean up remaining C library Tensor objects and
+        /// avoid a memory leak
+        /// </remarks>
+        public void Dispose()
+        {
+            if (_disposed)
+                return;
+            foreach (var runDetail in RunDetails)
+            {
+                foreach (var result in runDetail.Results)
+                {
+                    (result.Model as IDisposable)?.Dispose();
+                    result.IsModelDisposed = true;
+                }
```

Comment:
This really doesn't resolve the problem of one of the elements of the enumerable needing to have a different lifetime than the rest of the items.

Reply:
In that case, it makes sense to dispose of the entire `RunDetail runDetails` object instead of just the model, as I did just before [this commit](https://github.com/dotnet/machinelearning/pull/4978/commits/45681b424d09c03b2ed5c9764c8ee54eee8ce580). This is a bit inefficient as its disposing more than just the Model, but it ensures all the elements in the enumerable have the same lifetime.

Reply:
This class holds both an enumerable of IDisposable objects and a separate reference to one of the elements in that enumerable. This makes situation tricky because the requirement now is that the two elements have separate lifetimes. There isn't much difference between this implementation and the previous one because the BestRun model is a reference to an element in the enumerable and when that gets disposed, BestRun will too. 

This still doesn't meet @justinormont's main concern that the user might want to save BestRun and use it after the ExperimentResult object has been disposed. That use case is valid but it goes to counter to the common expectation that when an object is disposed, all its containing objects are disposed. The correct way to resolve this would require the following:

* BestRun is not a reference to an element in the enumerable but a separate copy. 
* When ExperimentResult / CrossValidationExperimentResult are diposed, they should dispose all the elements in the enumerable and BestRun
* But the user might want to manage memory more carefully. So they must be given an option to dispose the enumerable independently if they wish (without disposing the whole ExperimentResult / CrossValidationExperimentResult object)

cc: @justinormont 
Please let me know what you think. 




---
In reply to: [408451825](https://github.com/dotnet/machinelearning/pull/4978#discussion_r408451825) [](ancestors = 408451825)

Reply:
I believe the `BestRun` model is actually a separate copy and not a reference, as I explained why [here in my comment above](https://github.com/dotnet/machinelearning/pull/4978#discussion_r406985346). Please let me know if this isn't correct.